qurro/qarcoal.py

```diff
@@ -0,0 +1,113 @@
+
+    Parameters:
+    -----------
+        table: biom file with which to calculate log ratios
+        taxonomy: Qiime2 compliant feature taxonomy metadata file
+        num_string: numerator string to search for in taxonomy
+        denom_string: denominator string to search for in taxonomy
+        sample_to_use: Q2 Metadata file with samples to use (optional)
+
+    Returns:
+    --------
+        comparison_df: pd DataFrame in the form:
+
+            Sample-ID    Num_Sum    Denom_Sum   log_ratio
+                   S1          7           15   -0.762140
+    """
+
+    # biom table is features x samples
+    if samples_to_use is not None:
+        samp = set(samples_to_use.to_dataframe().index)
+        feat_table = table.filter(samp, axis="sample", inplace=False)
+        feat_table = feat_table.to_dataframe()
+    else:
+        feat_table = table.to_dataframe()
+
+    taxonomy_df = taxonomy.to_dataframe()
+
+    # taxonomy is features x [Taxon, Confidence]
+    taxonomy_df = taxonomy_df.loc[feat_table.index]
+    tax_num_df = taxonomy_df[taxonomy_df["Taxon"].str.contains(num_string)]
+    tax_denom_df = taxonomy_df[taxonomy_df["Taxon"].str.contains(denom_string)]
+
+    if tax_num_df.shape[0] == 0:
+        if tax_denom_df.shape[0] == 0:
+            raise (ValueError("neither feature found!"))
+        else:
+            raise (ValueError("numerator not found!"))
+    elif tax_denom_df.shape[0] == 0:
+        raise (ValueError("denominator not found!"))
+    else:
+        pass
+
+    # drop columns (samples) in which no feature w/ string is present
+    tax_num_df.drop(columns=["Taxon", "Confidence"], inplace=True)
+    tax_denom_df.drop(columns=["Taxon", "Confidence"], inplace=True)
+
+    tax_num_joined_df = tax_num_df.join(feat_table)
+    tax_num_joined_df = tax_num_joined_df.loc[
+        :, (tax_num_joined_df != 0).any(axis=0)
+    ]
+    tax_denom_joined_df = tax_denom_df.join(feat_table)
+    tax_denom_joined_df = tax_denom_joined_df.loc[
+        :, (tax_denom_joined_df != 0).any(axis=0)
+    ]
+
+    # keep only intersection of samples in which each feat string
+    # is present
+    samp_to_keep = set(tax_num_joined_df.columns).intersection(
+        set(tax_denom_joined_df.columns)
+    )
+    tax_num_joined_df = tax_num_joined_df[samp_to_keep]
+    tax_denom_joined_df = tax_denom_joined_df[samp_to_keep]
+
+    tax_num_sample_sum = tax_num_joined_df.sum(axis=0)
+    tax_denom_sample_sum = tax_denom_joined_df.sum(axis=0)
+
+    # sometimes 2 feature labels are the same i.e. same features
+    # don't want to check each time so only check if sums are
+    # the same
+    # e.g. if you are comparing g__A and s__B but g__A only
+    # appears when followed by s__B -> log ratios will all = 0
+    # TODO: Figure out if this makes sense at all lol
+    if tax_num_sample_sum.equals(tax_denom_sample_sum):
+        a = set(tax_num_df.index)
+        b = set(tax_num_df.index)
```

Comment:
I think this should be `set(tax_denom_df.index)`? As of now, if this block is reached then this will always fail with the ValueError.

Reply:
Oof... you're right. Serves me right for copy-pasting. Thought nothing of it since the test passed but that should be definitely be comparing num with denom indices.

Reply:
It's all good!

Reply:
Given the further discussion of the "same features" checking conditional, this is essentially resolved through the removal of this code block and the introduction of the `allow_shared_features` flag.